### DISCUSSION TITLE
Creating a variant that outputs two variations

### DISCUSSION THREAD

**Discussion by hteumeuleu** [tailwindlabs/tailwindcss#9017] [Help] [answered]
I’m trying to write a plugin for Tailwind to use in [Maizzle](https://maizzle.com/) (for HTML emails). I want to export dark mode styles the way I like it (again, for HTML emails, the default `dark` modifier is not enough.) I'm trying to declare a Variant that would let me add a `darkmode:bg-black` class. And it would export the following:

```css
@media (prefers-color-scheme:dark) {
	.darkmode-bg-black:not([class^="x_"]) {
		background-color: #000;
	}
}
[data-ogsb] .darkmode-bg-black {
	background-color: #000;
}
```

Here’s the code I have so far.

```js
const plugin = require('tailwindcss/plugin')
const darkmode = plugin(
	function({ addVariant, e, postcss }) {
      addVariant('darkmode', ({ container, separator }) => {
        const darkMediaRule = postcss.atRule({ name: 'media', params: '(prefers-color-scheme: dark)' })
        darkMediaRule.append(container.nodes)
        darkMediaRule.walkRules(rule => {
          rule.selector = `.${e(`darkmode${separator}`)}${rule.selector.slice(1)}${':not([class^="x_"])'}`
        })
        const ogsbRule = container
        ogsbRule.walkRules(rule => {
          rule.selector = `${'[data-ogsb]'} .${e(`darkmode${separator}`)}${rule.selector.slice(1)}`
        })
        container.append(darkMediaRule)
        container.append(ogsbRule)
      })
    }
)
module.exports = darkmode
```

This exports the first part correctly (with the `@media` rule), but not the second one. I understand this might be because rules inside the walkers are mutated, so the second walk here doesn’t find any match. So I’ve tried to play with `clone()` but I'm not getting anywhere. I feel like I’m close yet missing a big part to get it to work together. (Each walker work as I want them if I comment the other.)

Thanks in advance!

**Comment by wongjn** [comment]
Cloning should work:

```diff
 // …
 const darkMediaRule = postcss.atRule({ name: 'media', params: '(prefers-color-scheme: dark)' })
-darkMediaRule.append(container.nodes)
+darkMediaRule.append(container.clone().nodes)
 darkMediaRule.walkRules(rule => {
 // …
```
For the first pass with the at rule when you run `darkMediaRule.append()`, the nodes get moved from `container` to `darkMediaRule`, which is why `ogsbRule.walkRules()` doesn't generate anything. To fix this, we clone the container and use the nodes from the clone, so that `ogsbRule` will still have the original nodes to `walkRules()` through afterwards.

**Reply by hteumeuleu** [reply]
Thank you! This doesn't seem to work though. The output I get from this is a mix of both rules, without the `@media` query. 

```
[data-ogsb] .darkmode-bg-black:not([class^="x_"]) {
	background-color: #000 !important;
}
```

**Reply by wongjn** [reply]
I assumed you are working in Tailwind v2 (due to you using `-` as the variant separator), in which case this does work – [see the playground](https://play.tailwindcss.com/cg0kgibLAc).

**Comment by RobinMalfait** [comment, marked as answer]
Hey!

I think you will like the new [`addVariant` API](https://tailwindcss.com/docs/plugins#adding-variants) which makes this 100x easier.

```js
    function ({ addVariant }) {
      addVariant('darkmode', [
        '@media (prefers-color-scheme: dark) { &:not([class^="x_"]) }',
        '[data-ogsb] &',
      ])
    },
```
Example: https://play.tailwindcss.com/6WfjjZo3Nr

So given this input:
```html
<div class="darkmode:font-bold">Hello</div>
```
This is the generated output:
```css
@media (prefers-color-scheme: dark) {
  .darkmode\:font-bold:not([class^="x_"]) {
    font-weight: 700;
  }
}

[data-ogsb] .darkmode\:font-bold {
  font-weight: 700;
}
```

**Reply by hteumeuleu** [reply]
Thank you! This is a much cleaner approach and works as expected.

**Comment by arvigeus** [comment]
Is there a way to override default `dark` behavior? I need both versions, and currently have to generate both variants separately, then concatenate and minify them.

This does not work, unless I use custom name like `maybeDark`, but I don't want to modify existing files for this change

```js
function ({ addVariant }) {
    addVariant('dark', [
      '@media (prefers-color-scheme: dark)',
      '&.dark'
    ])
}
```

**Reply by cossssmin** [reply]
@arvigeus looks like setting `darkMode: 'whatever'` works, so in your case:

```js
const plugin = require('tailwindcss/plugin')

module.exports = {
  darkMode: 'nope',
  plugins: [
    plugin(function({ addVariant }) {
      addVariant('dark', [
      	'@media (prefers-color-scheme: dark)',
	    '&.dark'
      ])
    })
  ],
}
```

**Reply by arvigeus** [reply]
Just a heads up: this example will not work as expected: it will produce `.dark\:absolute.dark` instead of `.dark .dark\:absolute`.  
(I know you were copying my example, I just leave it for some poor future soul that might need the same as me)

In order to work correctly, just change the second argument to `.dark &`:

```js
module.exports = {
  darkMode: 'nope',
  plugins: [
    plugin(function({ addVariant }) {
      addVariant('dark', [
      	'@media (prefers-color-scheme: dark)',
	    '.dark &'
      ])
    })
  ],
}
```
